### ISSUE TITLE
Javascript Updates

### ISSUE THREAD

**Issue by Eric-Jackson** [mybb/mybb#3282]
Javascript needs some love, so I'm gonna work on that. All JS should also be moved from templates and put into JS files.

- [ ] general.js
- [ ] inline_edit.js
- [ ] inline_moderation.js
- [ ] inline_reports.js
- [ ] post.js
- [ ] question.js
- [ ] rating.js
- [ ] report.js
- [ ] thread.js
- [ ] usercp.js

**Comment by Eric-Jackson** [comment]
All collapsible sections will need a class of `collapse`, with whatever the toggle is having the class `collapse__toggle`, and the content being collapsed having a class of `collapse__content`. @justinsoltesz

**Comment by Eric-Jackson** [comment]
I'll toggle the class on the main container `collapse--collapsed` and `collapse--not-collapsed`, you can use those for styling the state.

**Comment by Shade-** [comment]
I guess if it's worth moving to TypeScript.

**Comment by Eric-Jackson** [comment]
@Shade- I don't think it is personally, since we're aiming to let people write theme exclusive Javascript. It'll make that a lot harder for most.

@justinsoltesz The ID from the collapse image currently used will need to be on the same element that has the `collapse` class.

**Comment by PaulBender** [comment]
Also, there is some hardcoded javascript in member.php (the inline validation for the registration form, look for $validator_javascript variable).

**Comment by Eric-Jackson** [comment]
Thanks @PaulBender , if you find any others just post them here!

**Comment by effone** [comment]
**Some hard-coded scripts found that can be driven separately**:

1. inc/functions.php: line 962 onwards: Some scattered hardcoded js chunk builder found in `redirect()` function.
2. usercp.php: line 3127, 3156, 3161
3. showthread.php: line 941
4. newreply.php: line 678
5. member.php: line 793

_(Findings under 'admin' & 'install' skipped.)_

**Templates containing scripts**:
- announcement
- calendar_addevent
- editpost
- forumdisplay_thread_rating
- forumjump_advanced
- global_dst_detection
- managegroup_adduser
- member_register_referrer
- member_register_regimage_recaptcha_invisible
- memberlist
- memberlist_search
- modcp_announcements_edit
- modcp_announcements_new
- modcp_banuser
- modcp_editprofile
- modcp_editprofile_signature
- modcp_finduser
- modcp_warninglogs
- moderation_delayedmoderation
- multipage_jump_page
- newreply
- newthread
- polls_editpoll
- polls_newpoll
- post_captcha_recaptcha_invisible
- private_advanced_search
- private_read
- private_send
- private_send_autocomplete
- report
- search
- showthread
- showthread_moderationoptions
- usercp_editlists
- warnings_warn
- warnings_warn_pm

**Comment by Eric-Jackson** [comment]
Awesome, thank you @effone ! The templates are on hold until @justinsoltesz  is finished with the theme, but the list will come in handy. :D

**Comment by Shade-** [comment]
What about using Webpack modules?

**Comment by Eric-Jackson** [comment]
@Shade- That requires a build step, so wouldn't be editable through the ACP or source files for people that aren't familiar with Node. I would love to use it for both Sass and JS, but that would require a big change to MyBB development requirements.

**Comment by Eric-Jackson** [comment]
I guess since there's a PHP build step, it wouldn't be that big of a deal to add a Node build step. The main problem would be editing through the ACP I believe.

**Comment by euantorano** [comment]
Yep, that’s the only problem. I want to make core JS editable in the ACP just like stylesheets.

**Comment by effone** [comment]
Huge thing what you have proposed @share but is our environment ready to adapt? Webpack can handle all the resources, not only js but completely config driven. I am not sure how we can handle that.

**Comment by Eric-Jackson** [comment]
@effone I had webpack working in my local 2.0 install, which had a Node build step, so we'll probably end up moving to it eventually. The environment is totally ready for it, it ultimately produces static CSS and JS anyway, just a matter of making it work without a Node dependency (So users can edit on shared hosting). Node doesn't work with shared hosting because all the users are sharing ports, and Node needs it's own port to operate on. This prevents us from using any of the hacks such as "install Node via PHP on shared hosting".

The one thing we need to solve to support Webpack, or any Node building is:
- Recompile source files after being edited in a shared hosting environment

Right now it will not be possible to get it working practically, no matter how useful it would be.

**Comment by Eric-Jackson** [comment]
I think it will be up to theme developers to offer Node building in their theme workflow.
